summingbird-online/src/main/scala/com/twitter/summingbird/online/executor/AsyncBase.scala

```diff
@@ -107,13 +136,15 @@ abstract class AsyncBase[I, O, S, D, RC](maxWaitingFutures: MaxWaitingFutures, m
     if (pendingFuturesCount > maxWaitingFuturesCount) {
       // Too many futures waiting, let's clear.
       outstandingFutures.dequeueAll(_.isDefined)
```

Comment:
that may work well. It is hard to guess often. Is there a real job you can test with to see how it performs or a benchmark suite?

We could add a `dropWhile` method to queue and on every call do `dropWhile(_.isDefined)` to very cheaply prune. When a future on the front gets stuck, this could trigger the `waitN` behavior, but as long as things were humming along, `dropWhile` might be a simple way to keep costs low.